src/ui/public/state_management/__tests__/hashing_store.js

```diff
@@ -0,0 +1,153 @@
+import expect from 'expect.js';
+import { delay as sleep } from 'bluebird';
+import { wrap } from 'lodash';
+import { encode as encodeRison } from 'rison-node';
+
+import StubBrowserStorage from 'test_utils/stub_browser_storage';
+import { HashingStore } from 'ui/state_management/hashing_store';
+
+const setup = ({ createHash } = {}) => {
+  const store = new StubBrowserStorage();
+  const hashingStore = new HashingStore({ store, createHash });
+  return { store, hashingStore };
+};
+
+describe('State Management Hashing Store', () => {
+  describe('#add', () => {
+    it('adds a value to the store and returns its hash', () => {
+      const { hashingStore, store } = setup();
+      const val = { foo: 'bar' };
+      const hash = hashingStore.add(val);
+      expect(hash).to.be.a('string');
+      expect(hash).to.be.ok();
+      expect(store._getValues()).to.have.length(1);
+    });
+
+    it('json encodes the values it stores', () => {
+      const { hashingStore, store } = setup();
+      const val = { toJSON() { return 1; } };
+      const hash = hashingStore.add(val);
+      expect(hashingStore.lookup(hash)).to.eql(1);
+    });
+
+    it('addresses values with a short hash', () => {
+      const val = { foo: 'bar' };
+      const longHash = 'longlonglonglonglonglonglonglonglonglonghash';
+      const { hashingStore } = setup({
+        createHash: () => longHash
+      });
+
+      const hash = hashingStore.add(val);
+      expect(hash.length < longHash.length).to.be.ok();
+    });
+
+    it('addresses values with a slightly longer hash when short hashes collide', () => {
+      const fixtures = [
+        {
+          hash: '1234567890-1',
+          val: { foo: 'bar' }
+        },
+        {
+          hash: '1234567890-2',
+          val: { foo: 'baz' }
+        },
+        {
+          hash: '1234567890-3',
+          val: { foo: 'boo' }
+        }
+      ];
+
+      const matchVal = json => f => JSON.stringify(f.val) === json;
+      const { hashingStore } = setup({
+        createHash: val => {
+          const fixture = fixtures.find(matchVal(val));
+          return fixture.hash;
+        }
+      });
+
+      const hash1 = hashingStore.add(fixtures[0].val);
+      const hash2 = hashingStore.add(fixtures[1].val);
+      const hash3 = hashingStore.add(fixtures[2].val);
+
+      expect(hash3.length - hash2.length).to.be(1);
```

Comment:
This would probably be clearer as `expect(hash3).to.have.length(hash2.length + 1)`